### PR TITLE
feat(ST-26): build student registration form

### PR DESCRIPTION
**Pull request by faysal-alam-sazim**
## Description of Change

- Added Registration module
- Built Student registration form
- Added validation
- Added a submit button shared component
- Added a link to the role choosing module

## Associated Ticket Link(s)

- https://trello.com/c/e3Y39D5q

## Related Pull Request(s)

- N/A

## Screenshots / Screen Recordings

https://github.com/user-attachments/assets/4bf4bb0f-9b3a-4e97-947a-ac1aa9a60022

Tablet View:
Screen Shot 2024-08-21 at 3 18 07 PM

Mobile View:
Screen Shot 2024-08-21 at 3 23 45 PM

Validation Check:

https://github.com/user-attachments/assets/f7342f5b-9f6f-462f-90c4-c1a45038c4e3

### CODE REVIEW

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+const StudentRegistrationForm = () => {
+  const { control, handleSubmit, watch } = useForm<TStudentRegistrationFormData>({
+    resolver: zodResolver(StudentRegistrationSchema),
+  });
+  const educationLevel = watch("educationLevel");
+
+  const onSubmit = (data: TStudentRegistrationFormData) => {
+    console.log(data);
+  };
```
Comment (ishmam-iqbal-sazim) [inline modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx:42]: Should be inside a hooks file and imported from there. (StudentRegistrationForm.hooks.ts)

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+  .refine((data) => data.password === data.confirmPassword, {
+    message: "Passwords don't match",
+    path: ["confirmPassword"],
+  });
+
+type TStudentRegistrationFormData = z.infer<typeof StudentRegistrationSchema>;
+
+const StudentRegistrationForm = () => {
+  const { control, handleSubmit, watch } = useForm<TStudentRegistrationFormData>({
+    resolver: zodResolver(StudentRegistrationSchema),
```
Comment (ishmam-iqbal-sazim) [inline modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx:36]:
```suggestion
    resolver: studentRegistrationSchemaResolver,
```


Put the validation schema inside a helpers file and import the validationSchemaResolver from that file.

Export would be something like this:
`export const studentRegistrationSchemaResolver = zodResolver(StudentRegistrationFormSchema)`

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.module.css [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/* StudentRegistrationForm.css */
```
Comment (ishmam-iqbal-sazim) [inline modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.module.css:1]: Let's remove unnecessary comments.

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.schema.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    degree: z.enum(["Bachelors", "Masters"]).optional(),
+    degreeName: z.string().optional(),
+    semesterYear: z.string().optional(),
+  })
+  .refine((data) => data.password === data.confirmPassword, {
+    message: "Passwords don't match",
+    path: ["confirmPassword"],
+  });
+
+export type TStudentRegistrationFormData = z.infer<typeof StudentRegistrationSchema>;
```
Comment (ishmam-iqbal-sazim) [inline modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.schema.ts:26]: Let's be explicit with the types.

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.module.css [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+@media (max-width: 768px) {
+  .form-container {
+    width: 100%;
+  }
+
+  .input-flex {
+    flex-direction: column;
+  }
```
Comment (ragib-hossain-sazim) [inline modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.module.css:24]: [NITPICK] add newline

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.schema.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    firstName: z.string().min(1, "First Name is required"),
+    lastName: z.string().min(1, "Last Name is required"),
+    gender: z.enum(["Male", "Female"]),
+    address: z.string().min(1, "Address is required"),
+    phoneNumber: z.string().min(1, "Phone Number is required"),
+    email: z.string().email("Invalid email address"),
+    password: z.string().min(6, "Password must be at least 6 characters"),
+    confirmPassword: z.string().min(6, "Confirm Password is required"),
+    educationLevel: z.enum(["School", "College", "University"]),
+    medium: z.enum(["Bangla", "English"]).optional(),
```
Comment (ragib-hossain-sazim) [inline modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.schema.ts:15]: [JC] does the requirement specify anywhere that these fields should be optional? These are mandatory fields.

File: pages/register/student/index.page.tsx [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import React from "react";
+
+import StudentRegistrationContainer from "@/modules/Register/containers/StudentRegistrationContainer";
+
+const StudentRegiseterPage = () => <StudentRegistrationContainer />;
```
Comment (ragib-hossain-sazim) [inline pages/register/student/index.page.tsx:5]: `StudentRegiseterPage` => `StudentRegistrationPage`

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+                        styles={{
+                          label: { color: "green" },
+                        }}
+                      />
+                    )}
+                  />
+                ) : null}
+              </Box>
+
+              {educationLevel === "School" || educationLevel === "College" ? (
```
Comment (ragib-hossain-sazim) [inline modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx:172]: create a constants file in shared folder and define these Education levels as constants.

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+                  <Controller
+                    name="degreeName"
+                    control={control}
+                    render={({ field }) => (
+                      <Select
+                        {...field}
+                        label="Degree Name"
+                        required
+                        data={[
+                          "Bachelor of Science",
```
Comment (ragib-hossain-sazim) [inline modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx:202]: these options are fine, no need change it.

File: modules/Register/hooks/StudentRegistrationForm.hooks.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import { TStudentRegistrationFormData } from "../components/StudentRegistrationForm/StudentRegistrationForm.schema";
+
+const onSubmit = (data: TStudentRegistrationFormData) => {
+  console.log(data);
```
Comment (ragib-hossain-sazim) [inline modules/Register/hooks/StudentRegistrationForm.hooks.ts:4]: remove unnecessary console logs.

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx (hunk 8 above)
Comment (faysal-alam-sazim) [reply modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx:172]: Created enum file for this.

File: modules/Register/hooks/StudentRegistrationForm.hooks.ts (hunk 10 above)
Comment (faysal-alam-sazim) [reply modules/Register/hooks/StudentRegistrationForm.hooks.ts:4]: Will remove it when implementing API call

Comment (ishmam-iqbal-sazim) [reply modules/Register/hooks/StudentRegistrationForm.hooks.ts:4]: Add // TODO comment with a message to mark these for later removal

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+                            "Class 3",
+                            "Class 4",
+                            "Class 5",
+                            "Class 7",
+                            "Class 8",
+                            "Class 9",
+                            "Class 10",
+                          ]
+                        : ["Class 11", "Class 12"]
+                    }
```
Comment (ishmam-iqbal-sazim) [inline modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.tsx:197]: Put these arrays in a constants file `StudentRegistrationForm.constants.ts` and export from there.

File: modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.schema.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        });
+      }
+      if (!data.semesterYear) {
+        ctx.addIssue({
+          code: z.ZodIssueCode.custom,
+          message: "Semester/Year is required",
+          path: ["semesterYear"],
+        });
+      }
+    }
```
Comment (ishmam-iqbal-sazim) [inline modules/Register/components/StudentRegistrationForm/StudentRegistrationForm.schema.ts:74]: Nice, this looks great!

Would be helpful if you could put screen recordings in the PR description to showcase this validation.

**Review by ragib-hossain-sazim — APPROVED** [review]

**Review by ishmam-iqbal-sazim — APPROVED** [review]

**Pull request merged**